wallet/src/de/schildbach/wallet/ui/dashpay/PlatformRepo.kt

```diff
@@ -142,9 +150,66 @@ class PlatformRepo(val walletApplication: WalletApplication) {
     suspend fun isNameRegisteredAsync(blockchainIdentity: BlockchainIdentity) {
         withContext(Dispatchers.IO) {
             val (result, usernames) = blockchainIdentity.watchUsernames(blockchainIdentity.getUsernamesWithStatus(BlockchainIdentity.UsernameStatus.REGISTRATION_PENDING), 10, 5000, BlockchainIdentity.RetryDelayType.SLOW20)
-            if(!result) {
+            if (!result) {
                 throw TimeoutException("the usernames: $usernames were not found to be registered in the allotted amount of time")
             }
         }
     }
+
+    suspend fun initBlockchainIdentityData(username: String): BlockchainIdentityData {
+        return blockchainIdentityDataDaoAsync.load()
+                ?: BlockchainIdentityData(BlockchainIdentityData.State.UPGRADING_WALLET, false, username)
+    }
+
+    fun initBlockchainIdentity(blockchainIdentityData: BlockchainIdentityData, wallet: Wallet): BlockchainIdentity {
+        if (blockchainIdentityData.creditFundingTxId != null) {
+            val creditFundingTx = wallet.getTransaction(blockchainIdentityData.creditFundingTxId)
+            if (creditFundingTx != null) {
+                val creditFundingTransaction = wallet.getCreditFundingTransaction(creditFundingTx)
+                return BlockchainIdentity(Identity.IdentityType.USER, creditFundingTransaction, wallet).apply {
+                    currentUsername = blockchainIdentityData.username
+                    // should we load `preorderSalt` somehow?
+                    registrationStatus = blockchainIdentityData.registrationStatus!!
+                    // should we load `usernameStatus` somehow?
```

Comment:
Yes.

```
var usernameStatus = HashMap<String, Any>()
usernameStatus[BLOCKCHAIN_USERNAME_STATUS] = blockchainIdentityData.usernameStatus
usernameStatus[BLOCKCHAIN_USERNAME_SALT] = blockchainIdentityData.preorderSalt
usernameStatuses[username] = usernameStatus
```

There may be a cooler way to do the code in Kotlin.

Reply:
Maybe later we can do a data class for the username status instead of these maps.